snforge_std/src/event.cairo

```diff
@@ -0,0 +1,206 @@
+use array::ArrayTrait;
+use array::SpanTrait;
+use clone::Clone;
+use serde::Serde;
+use option::OptionTrait;
+use traits::Into;
+
+use starknet::testing::cheatcode;
+use starknet::ContractAddress;
+
+#[derive(Drop, Serde)]
+enum SpyOn {
+    All: (),
+    One: ContractAddress,
+    Multiple: Array<ContractAddress>
+}
+
+fn spy_events(spy_on: SpyOn) -> EventSpy {
+    let mut inputs = array![];
+    spy_on.serialize(ref inputs);
+    let output = cheatcode::<'spy_events'>(inputs.span());
+
+    EventSpy { _id: *output[0], events: array![] }
+}
+
+fn event_name_hash(name: felt252) -> felt252 {
+    let mut output = cheatcode::<'event_name_hash'>(array![name].span());
+    *output[0]
+}
+
+#[derive(Drop, Clone, Serde)]
+struct Event {
+    from: ContractAddress,
+    name: felt252,
+    keys: Array<felt252>,
+    data: Array<felt252>
+}
+
+#[derive(Drop, Clone, Serde)]
+struct EventSpy {
+    _id: felt252,
+    events: Array<Event>,
+}
+
+trait EventFetcher {
+    fn fetch_events(ref self: EventSpy);
+}
+
+impl EventFetcherImpl of EventFetcher {
+    fn fetch_events(ref self: EventSpy) {
```

Comment:
What if I call this, and then call another spied contract, and then fetch again? Will it fetch duplicates?

Reply:
Nope, we do not have one but you're right. It won't fetch duplicates